### DISCUSSION TITLE
Random solutions

### DISCUSSION THREAD

**Discussion by Ahmed-Emad10** [cvc5/cvc5#9581] [Q&A] [answered]
Hello there, 
I wanted to ask if CVC5 generates solution randomly or the first solution it finds?
also, if I run the program many times on the same problem will it generate many solutions or the same solution?
because I noticed that in another tool the solution it finds is the first solution it finds and if I run that tool many times in the same problem it gives me the same results.
Thanks in advance.

**Comment by martin-cs** [comment, marked as answer]
cvc5 should be deterministic.  For the same input you should get the same output.  This is a design goal because it makes it possible to test and means it is more reliable.

However, note that "the same" is quite a precise statement.  If you re-order the asserts or change expressions then you may well get  different results.

There have been tools written to do uniform random sampling of solutions to SAT problems.  I am not sure of the current state of tools to do random sampling of SMT solutions.

**Reply by martin-cs** [reply]
Oh, one quick follow up.  When I have asked people about uniform random sampling of SMT solutions, a really key issue is what theory you are using.  If it has a finite number of models then it's conceptually straight-forward.  If there are a potentially infinite number then the conversation seems to get side-tracked by questions of measure theory and if probability is even a meaningful concept.

**Reply by barrettcw** [reply]
Here's a nice approach for "random" sampling of SMT solutions: https://ieeexplore.ieee.org/abstract/document/8587695
